[Flaky Test] go_test: //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2

<!-- Please only use this template for submitting reports about flaky tests or jobs (pass or fail with no underlying change in code) in Kubernetes CI -->

**Which jobs are flaking**:
bazel-test-master (periodic-kubernetes-bazel-test-master #1314084281143791616)

**Which test(s) are flaking**:
go_test: //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2 

**Testgrid link**:
https://testgrid.k8s.io/google-unit#bazel-test-master
**Reason for failure**:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/periodic-kubernetes-bazel-test-master/1314084281143791616

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //staging/src/k8s.io/apiserver/pkg/storage/cacher:go_default_test
-----------------------------------------------------------------------------
W1008 06:36:28.867515  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:34.958950  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:36.960954  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:39.179064  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:42.651246  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
--- FAIL: TestDispatchEventWillNotBeBlockedByTimedOutWatcher (0.11s)
    cacher_whitebox_test.go:1072: watcher is blocked by slower one (count: 22)
W1008 06:36:43.498316  209647 reflector.go:436] k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go:503: watch of *v1.Pod ended with: very short watch: k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go:503: Unexpected watch close - watch lasted less than a second and no items received
FAIL
```

**Anything else we need to know**:
- links to go.k8s.io/triage appreciated

https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=apiserver%2Fpkg%2Fstorage%2Fcacher%2Fgo_default_test

```


Time | Job | Test
-- | -- | --
08/10/2020, 11:36:02 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
07/10/2020, 20:49:41 | periodic-kubernetes-bazel-test-canary | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
07/10/2020, 03:49:04 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
02/10/2020, 13:06:56 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
02/10/2020, 10:40:48 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2


```

- links to specific failures in spyglass appreciated

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/periodic-kubernetes-bazel-test-master/1314084281143791616

<!-- Please see the deflaking doc (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/flaky-tests.md) for more guidance! -->

cc @kubernetes/ci-signal

--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
@@ -263,7 +263,7 @@ type Cacher struct {
 
 	// Defines a time budget that can be spend on waiting for not-ready watchers
 	// while dispatching event before shutting them down.
-	dispatchTimeoutBudget *timeBudget
+	dispatchTimeoutBudget timeBudget
 
 	// Handling graceful termination.
 	stopLock sync.RWMutex

--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go
@@ -998,6 +998,14 @@ func TestBookmarksOnResourceVersionUpdates(t *testing.T) {
 	wg.Wait()
 }
 
+type fakeTimeBudget struct{}
+
+func (f *fakeTimeBudget) takeAvailable() time.Duration {
+	return 2 * time.Second
+}
+
+func (f *fakeTimeBudget) returnUnused(_ time.Duration) {}
+
 func TestDispatchEventWillNotBeBlockedByTimedOutWatcher(t *testing.T) {
 	backingStorage := &dummyStorage{}
 	cacher, _, err := newTestCacher(backingStorage)
@@ -1010,7 +1018,19 @@ func TestDispatchEventWillNotBeBlockedByTimedOutWatcher(t *testing.T) {
 	cacher.ready.wait()
 
 	// Ensure there is some budget for slowing down processing.
-	cacher.dispatchTimeoutBudget.returnUnused(50 * time.Millisecond)
+	// When using the official `timeBudgetImpl` we were observing flakiness
+	// due under the following conditions:
+	// 1) the watch w1 is blocked, so we were consuming the whole budget once
+	//    its buffer was filled in (10 items)
+	// 2) the budget is refreshed once per second, so it basically wasn't
+	//    happening in the test at all
+	// 3) if the test was cpu-starved and we weren't able to consume events
+	//    from w2 ResultCh it could have happened that its buffer was also
+	//    filling in and given we no longer had timeBudget (consumed in (1))
+	//    trying to put next item was simply breaking the watch
+	// Using fakeTimeBudget gives us always a budget to wait and have a test
+	// pick up something from ResultCh in the meantime.
+	cacher.dispatchTimeoutBudget = &fakeTimeBudget{}
 
 	makePod := func(i int) *examplev1.Pod {
 		return &examplev1.Pod{
@@ -1055,8 +1075,6 @@ func TestDispatchEventWillNotBeBlockedByTimedOutWatcher(t *testing.T) {
 				shouldContinue = false
 				break
 			}
-			// Ensure there is some budget for fast watcher after slower one is blocked.
-			cacher.dispatchTimeoutBudget.returnUnused(50 * time.Millisecond)
 			if event.Type == watch.Added {
 				eventsCount++
 				if eventsCount == totalPods {

--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go
@@ -39,7 +39,12 @@ const (
 // NOTE: It's not recommended to be used concurrently from multiple threads -
 // if first user takes the whole timeout, the second one will get 0 timeout
 // even though the first one may return something later.
-type timeBudget struct {
+type timeBudget interface {
+	takeAvailable() time.Duration
+	returnUnused(unused time.Duration)
+}
+
+type timeBudgetImpl struct {
 	sync.Mutex
 	budget time.Duration
 
@@ -47,8 +52,8 @@ type timeBudget struct {
 	maxBudget time.Duration
 }
 
-func newTimeBudget(stopCh <-chan struct{}) *timeBudget {
-	result := &timeBudget{
+func newTimeBudget(stopCh <-chan struct{}) timeBudget {
+	result := &timeBudgetImpl{
 		budget:    time.Duration(0),
 		refresh:   refreshPerSecond,
 		maxBudget: maxBudget,
@@ -57,7 +62,7 @@ func newTimeBudget(stopCh <-chan struct{}) *timeBudget {
 	return result
 }
 
-func (t *timeBudget) periodicallyRefresh(stopCh <-chan struct{}) {
+func (t *timeBudgetImpl) periodicallyRefresh(stopCh <-chan struct{}) {
 	ticker := time.NewTicker(time.Second)
 	defer ticker.Stop()
 	for {
@@ -74,7 +79,7 @@ func (t *timeBudget) periodicallyRefresh(stopCh <-chan struct{}) {
 	}
 }
 
-func (t *timeBudget) takeAvailable() time.Duration {
+func (t *timeBudgetImpl) takeAvailable() time.Duration {
 	t.Lock()
 	defer t.Unlock()
 	result := t.budget
@@ -82,7 +87,7 @@ func (t *timeBudget) takeAvailable() time.Duration {
 	return result
 }
 
-func (t *timeBudget) returnUnused(unused time.Duration) {
+func (t *timeBudgetImpl) returnUnused(unused time.Duration) {
 	t.Lock()
 	defer t.Unlock()
 	if unused < 0 {

--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget_test.go
@@ -22,7 +22,7 @@ import (
 )
 
 func TestTimeBudget(t *testing.T) {
-	budget := &timeBudget{
+	budget := &timeBudgetImpl{
 		budget:    time.Duration(0),
 		maxBudget: time.Duration(200),
 	}

